common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java

```diff
@@ -14,91 +14,132 @@
+        List<Pair<CustomColor, Float>> circles = circlesToRender.get(player);
+        if (circles == null) return;
 
-        renderCircleWithRadius(e.getPoseStack(), 8, e.getPlayer().position(), circleColors.get(e.getPlayer()));
+        circles.forEach(circleType -> {
+            float radius = circleType.b();
+            int color = circleType.a().asInt();
+
+            renderCircle(e.getPoseStack(), player.position(), radius, color);
+        });
     }
 
     @SubscribeEvent
     public void onTick(TickEvent e) {
-        playerIsRendered.forEach((player, rendered) -> {
-            CustomColor color = getCircleColor(player);
-            if (rendered && color != null) {
-                circleColors.put(player, color);
-            } else {
-                circleColors.remove(player);
-            }
-        });
-        playerIsRendered.clear();
+        // Once every tick, calculate circles for all players that were detected (i.e. rendered)
+        // since the previous tick
+        circlesToRender.clear();
+        detectedPlayers.forEach(this::checkMajorIdCircles);
+        detectedPlayers.clear();
     }
 
-    private CustomColor getCircleColor(Player player) {
-        if (!Models.Player.isLocalPlayer(player)) return null; // Don't render for ghost/npc
-        String playerName = ComponentUtils.getUnformatted(player.getName());
-        boolean isSelf =
-                ComponentUtils.getUnformatted(McUtils.player().getName()).equals(playerName);
-        if (isSelf && McUtils.mc().screen instanceof InventoryScreen)
-            return null; // Don't render for preview in inventory
-        if (!Models.Party.getPartyMembers().contains(playerName) && !isSelf)
-            return null; // Other players must be in party
-
-        // We are getting the item info the same way as GearViewerScreen since we care about other people's items
-        String gearName = ComponentUtils.getUnformatted(player.getMainHandItem().getHoverName());
-        GearInfo gearInfo = Models.Gear.getGearInfoFromApiName(gearName);
-        if (gearInfo == null) return null;
-
-        if (isSelf) { // Do not render if the item is not for the player's class
-            if (gearInfo.requirements().classType().isEmpty()) return null;
-            ClassType classType = gearInfo.requirements().classType().get();
-            if (classType != Models.Character.getClassType()) return null;
+    private void checkMajorIdCircles(Player player) {
+        // Don't render for ghost/npc
+        if (!Models.Player.isLocalPlayer(player)) return;
+
+        List<GearInfo> validGear;
+
+        if (player == McUtils.player()) {
+            // This is ourselves, rendered from outside
+
+            // Don't render for preview in inventory or character selection screen
+            if (McUtils.mc().screen instanceof InventoryScreen) return;
+            if (McUtils.mc().screen instanceof CharacterSelectorScreen) return;
```

Comment:
Check for the wynntils book screen as well.

Reply:
Good point, I will.